beetsplug/lyrics.py

```diff
@@ -209,105 +193,151 @@ def slug(text):
-        return urllib.parse.quote(s)
+    def fetch(self, artist, title, album=None, length=None):
+        raise NotImplementedError
 
-    def build_url(self, artist, title):
-        return self.URL_PATTERN % (
-            self._encode(artist.title()),
-            self._encode(title.title()),
-        )
 
-    def fetch_url(self, url):
-        """Retrieve the content at a given URL, or return None if the source
-        is unreachable.
-        """
-        try:
-            # Disable the InsecureRequestWarning that comes from using
-            # `verify=false`.
-            # https://github.com/kennethreitz/requests/issues/2214
-            # We're not overly worried about the NSA MITMing our lyrics scraper
-            with warnings.catch_warnings():
-                warnings.simplefilter("ignore")
-                r = requests.get(
-                    url,
-                    verify=False,
-                    headers={
-                        "User-Agent": USER_AGENT,
-                    },
-                    timeout=10,
-                )
-        except requests.RequestException as exc:
-            self._log.debug("lyrics request failed: {0}", exc)
-            return
-        if r.status_code == requests.codes.ok:
-            return r.text
-        else:
-            self._log.debug("failed to fetch: {0} ({1})", url, r.status_code)
-            return None
+class LRCLibItem(TypedDict):
+    """Definition of a single lyrics JSON object returned by the LRCLib API."""
 
-    def fetch(self, artist, title, album=None, length=None):
-        raise NotImplementedError()
+    id: int
+    name: str
+    trackName: str
+    artistName: str
+    albumName: str
+    duration: float
+    instrumental: bool
+    plainLyrics: str
+    syncedLyrics: str | None
 
 
 class LRCLib(Backend):
-    base_url = "https://lrclib.net/api/get"
+    base_url = "https://lrclib.net/api/search"
 
-    def fetch(self, artist, title, album=None, length=None):
+    @staticmethod
+    def get_rank(
+        target_duration: float, litem: LRCLibItem
+    ) -> tuple[float, bool]:
+        """Rank the given lyrics item.
+
+        Return a tuple with the following values:
+        1. Difference between item lyrics duration and the item duration in sec
+        2. Boolean telling whether synced lyrics are available.
+        """
+        return (
+            abs(litem["duration"] - target_duration),
+            not litem["syncedLyrics"],
+        )
+
+    @classmethod
+    def pick_lyrics(
+        cls, target_duration: float, data: list[LRCLibItem]
+    ) -> LRCLibItem:
+        """Return best matching lyrics item from the given list.
+
+        Note that the incoming list is guaranteed to be non-empty.
+        """
+        return min(data, key=lambda item: cls.get_rank(target_duration, item))
+
+    def fetch(
+        self,
+        artist: str,
+        title: str,
+        album: str | None = None,
+        length: float = 0.0,
```

Comment:
Just to be clear, does this mean that the default ranking will prioritize search results with shorter durations?  I don't know whether that's a good idea.

Reply:
Nope, it prioritizes lyrics with duration closest to the original item duration 
```python
        return (
            abs(litem["duration"] - target_duration),
            not litem["syncedLyrics"],
        )
```

Clarified this in the docstrings